### PR TITLE
#7078: This is a refactor to better handle optional output tensors

### PR DESCRIPTION
**Pull request by eyonland**
A request to support optional tensors in create_program was originally made on https://github.com/tenstorrent-metal/tt-metal/issues/5163 . It became clear that support should be provided for cases were optional tensors could be ignored.

The approach taken here was to allow for std::vector<std::optional<Tensor>> to be provided for the output tensors in create_program which are always the last argument.   The expectation is that the return type of create_output_tensors would also match the output_tensors of create_program.

### CODE REVIEW

File: tt_metal/tools/profiler/op_profiler.hpp [hunk 1]
```diff
@@ -166,12 +166,26 @@ namespace op_profiler {
         return ret;
     }
 
+    static inline vector<json> get_tensors_json(const vector<std::optional<Tensor>>& tensors)
```
Comment (mo-tenstorrent) [inline tt_metal/tools/profiler/op_profiler.hpp:169]: the `const Tensor` version on line 155 could not be used?

**Review by arakhmati — APPROVED** [review]

File: tt_metal/tools/profiler/op_profiler.hpp (hunk 1 above)
Comment (eyonland) [reply tt_metal/tools/profiler/op_profiler.hpp:169]: That was for std::optional<const Tensor>>& tensors and not std::optional<Tensor>>& tensors.  Notice the const in front of Tensors.  I can put more template magic here and reuse the same function if you think that would be better.

**Review by arakhmati — APPROVED** [review]

File: tt_metal/tools/profiler/op_profiler.hpp [hunk 2]
```diff
@@ -235,26 +249,28 @@ namespace op_profiler {
...
     inline std::string op_meta_data_serialized_json(
             uint32_t opID,
             uint32_t device_id,
-            const tt::tt_metal::operation::DeviceOperation& op,
+            const DeviceOperationType<OutputTensors>& op,
             const std::variant<std::shared_ptr<Program>, std::reference_wrapper<Program>>& program,
             const std::vector<Tensor>& input_tensors,
             const std::vector<std::optional<const tt::tt_metal::Tensor>>& optional_input_tensors,
-            std::vector<Tensor>& output_tensors)
+            OutputTensors& output_tensors)
```
Comment (mo-tenstorrent) [inline tt_metal/tools/profiler/op_profiler.hpp:273]: CI right now does not do a through check on the generated op reports. Did you have a chance to sanity check an op-report and its input output tensor columns.

File: tt_eager/tt_dnn/op_library/run_operation.cpp [hunk 3]
```diff
@@ -99,45 +122,50 @@ std::vector<Tensor> run_host_operation(const HostOperation& operation, const std
...
             }
-            auto& program_with_callbacks = *(reinterpret_cast<operation::ProgramWithCallbacks*>(program_ptr.value().get()));
+            auto& program_with_callbacks = *(reinterpret_cast<operation::CacheableProgram<OutputTensors>*>(program_ptr.value().get()));
             TT_ASSERT(program_with_callbacks.supports_program_cache());
 
             if (cache_hit) {
                 ZoneScopedN("Cache_hit_set_runtime_args");
                 if (program_with_callbacks.override_addresses_callback.has_value()) {
                     auto override_addresses_callback = program_with_callbacks.override_addresses_callback.value();
+                    // Deprecated
```
Comment (tt-asaigal) [inline tt_eager/tt_dnn/op_library/run_operation.cpp:168]: Hey Eyon, maybe I'm missing something but why is this API deprecated? We have a template specialization that allows output tensors of type `Tensors&`.

**Review by tt-asaigal — APPROVED** [review]

**Review by mo-tenstorrent — APPROVED** [review]

File: tt_metal/tools/profiler/op_profiler.hpp (hunk 2 above)
Comment (eyonland) [reply tt_metal/tools/profiler/op_profiler.hpp:273]: It looks good to me.

File: tt_metal/tools/profiler/profile_this.py [hunk 4]
```diff
@@ -26,7 +26,7 @@ def profile_command(test_command, output_folder, name_append):
         options += f"-o {output_folder}"
     if name_append:
         options += f" -n {name_append}"
-    opProfilerTestCommand = f"python -m tracy -r -p {options} -m {test_command}"
+    opProfilerTestCommand = f"python -m tracy -v -r -p {options} -m {test_command}"
```
Comment (eyonland) [inline tt_metal/tools/profiler/profile_this.py:29]: @mo-tenstorrent , are you ok with adding "-v" here by default?

Comment (mo-tenstorrent) [reply tt_metal/tools/profiler/profile_this.py:29]: Let's see if folks will start complaining about their run logs getting crowded 😄

**Pull request merged**